src/tables/files/files.windows.cc

```diff
@@ -0,0 +1,318 @@
+#include "util/windows-util.h"
+
+#include <regex>
+
+#include <Shlwapi.h>
+
+namespace zeek::agent::table {
+
+class FilesListWindows final : public FilesListCommon {
+public:
+    std::vector<std::vector<Value>> snapshot(const std::vector<table::Argument>& args) override;
+
+private:
+    std::vector<Value> buildFileRow(const std::string& pattern, const WIN32_FIND_DATAA& data) const;
+};
+
+class FilesLinesWindows final : public FilesLinesCommon {
+public:
+    std::vector<std::vector<Value>> snapshot(const std::vector<table::Argument>& args) override;
+};
+
+class FilesColumnsWindows final : public FilesColumnsCommon {
+public:
+    std::vector<std::vector<Value>> snapshot(const std::vector<table::Argument>& args) override;
+};
+
+namespace {
+database::RegisterTable<FilesListWindows> _1;
+database::RegisterTable<FilesLinesWindows> _2;
+database::RegisterTable<FilesColumnsWindows> _3;
+} // namespace
+
+static std::string get_full_path(const std::string& pattern, const char* filename) {
+    char ret[MAX_PATH];
+    char* tmp = new char[pattern.size() + 1];
+    strncpy(tmp, pattern.c_str(), pattern.size());
+    tmp[pattern.size()] = '\0';
+
+    PathRemoveFileSpecA(tmp);
+    PathCombineA(ret, tmp, filename);
+    delete[] tmp;
+    return ret;
+}
+
+std::pair<std::string, std::vector<filesystem::path>> FilesBase::expandPaths(const std::vector<table::Argument>& args) {
+    logger()->error("FilesBase::expandPaths is not implemented on Windows");
+    return {};
+}
+
+std::vector<Value> FilesListWindows::buildFileRow(const std::string& pattern, const WIN32_FIND_DATAA& data) const {
+    std::vector<Value> row;
+    std::string full_path = get_full_path(pattern, data.cFileName);
+
+    Value path = full_path;
+    Value type;
+
+    uint16_t temp_mode = 0;
+    if ( data.dwFileAttributes & FILE_ATTRIBUTE_READONLY )
+        temp_mode = 0444;
+    else
+        temp_mode = 0666;
+
+    if ( data.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY ) {
+        // TODO: not sure this addition to the mode is correct
+        temp_mode += 0111;
+        type = "dir";
+    }
+    else {
+        // The data from FindFirstFile doesn't include a file type and the handle that's
+        // returned isn't a file handle (it's a search handle). Open the file separately
+        // and call GetFileType(), and then close it again.
+        DWORD file_type = FILE_TYPE_UNKNOWN;
+        HANDLE file_handle = CreateFileA(full_path.c_str(), GENERIC_READ, 0, NULL, OPEN_EXISTING, 0, NULL);
+
+        if ( file_handle != INVALID_HANDLE_VALUE ) {
+            file_type = GetFileType(file_handle);
+            if ( file_type == FILE_TYPE_UNKNOWN && GetLastError() != NO_ERROR ) {
+                std::error_condition cond =
+                    std::system_category().default_error_condition(static_cast<int>(GetLastError()));
+                logger()->error(
+                    format("Failed to get file type for {}: {} ({})", full_path, cond.message(), cond.value()));
```

Comment:
Generally, prefer `warn` over `error` for tables running into unexpected situations. And report only if it's really unexpected (can't tell the semantics here and elsewhere)